src/client/opamClient.ml

```diff
@@ -202,9 +202,25 @@ let upgrade_t
         else OpamPackage.packages_of_names t.installed requested
       in
       let latest =
-        OpamPackage.Name.Set.fold (fun name acc ->
-            OpamPackage.Set.add (OpamPackage.max_version t.packages name) acc)
-          (OpamPackage.names_of_packages to_check)
+        OpamPackage.Set.fold (fun pkg acc ->
+            let name = OpamPackage.name pkg in
+            let pkgs = OpamPackage.packages_of_name t.packages name in
```

Comment:
I don't know if it worth working on the set of version instead of package one.

Reply:
it simplifies the code greatly. Otherwise we would need to have an intermediate map of set and update it